File: src/adapter.rs
use rusb::constants::LIBUSB_DT_HID;
use rusb::{DeviceHandle, Direction, GlobalContext};
use std::time::Duration;

const MAIN_STICK: StickRange = StickRange::new(0x80, 0x80, 0x7F);
const C_STICK: StickRange = StickRange::new(0x80, 0x80, 0x7F);

const PAYLOAD_SIZE: usize = 37;
const ALLOWED_TIMEOUT: Duration = Duration::from_millis(16);

#[derive(Debug)]
pub enum Error {
    Rusb(rusb::Error),
    Adapter,
    Payload,
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum Port {
    One,
    Two,
    Three,
    Four,
}

impl Port {
    pub fn channel(self) -> usize {
        match self {
            Port::One => 0,
            Port::Two => 1,
            Port::Three => 2,
            Port::Four => 3,
        }
    }

    pub fn all() -> &'static [Port] {
        const PORTS: [Port; 4] = [Port::One, Port::Two, Port::Three, Port::Four];
        &PORTS
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum Rumble {
    Off,
    On,
}

impl From<Rumble> for u8 {
    fn from(state: Rumble) -> u8 {
        match state {
            Rumble::Off => 0,
            Rumble::On => 1,
        }
    }
}

pub struct Adapter {
    handle: DeviceHandle<GlobalContext>,
    endpoint_in: u8,
    endpoint_out: u8,
    /// `None` if the controller on the channel is unplugged. If it is plugged,
    /// the drift is calculated once when connected on `read_inputs`.
    drifts: [Option<Drift>; 4],
}

impl Adapter {
    pub fn open() -> Result<Adapter, Error> {
        const GC_VID: u16 = 0x057E;
        const GC_PID: u16 = 0x0337;

        let mut handle = rusb::open_device_with_vid_pid(GC_VID, GC_PID).ok_or(Error::Adapter)?;

        handle.claim_interface(0).map_err(Error::Rusb)?;

        let (endpoint_in, endpoint_out) = Self::endpoints(&handle)?;

        // From Dolphin:
        // This call makes Nyko-brand (and perhaps other) adapters work.
        // However it returns LIBUSB_ERROR_PIPE with Mayflash adapters.
        let _err = handle.write_control(0x21, 11, 0x0001, 0, &[], Duration::from_secs(1));

        // Not sure what this does but Dolphin does it
        let _ = handle
            .write_interrupt(endpoint_out, &[0x13], ALLOWED_TIMEOUT)
            .map_err(Error::Rusb)?;

        let adapter = Adapter {
            handle,
            endpoint_in,
            endpoint_out,
            drifts: Default::default(),
        };

        adapter.reset_rumble()?;

        Ok(adapter)
    }

    pub fn close(&mut self) -> Result<(), Error> {
        self.handle.release_interface(0).map_err(Error::Rusb)
    }

    pub fn read_inputs(&mut self) -> Result<[Option<Input>; 4], Error> {
        let payload = self.read_payload()?;
        let mut inputs: [Option<Input>; 4] = Default::default();

        for port in Port::all() {
            let chan = port.channel();
            // type is 0 if no controller is plugged, 1 if wired, and 2 if wireless
            let r#type = payload[1 + (9 * chan)] >> 4;
            let connected = r#type != 0;

            if !connected {
                self.drifts[chan] = None;
                continue;
            }

            let raw = Input::parse(&payload, *port);
            let drift = self.drifts[chan].get_or_insert_with(|| Drift::new(&raw));

            inputs[chan] = Some(drift.correct(raw));
        }

        Ok(inputs)
    }

    pub fn set_rumble(&self, states: [Rumble; 4]) -> Result<(), Error> {
        let payload = [
            0x11,
            states[0].into(),
            states[1].into(),
            states[2].into(),
            states[3].into(),
        ];

        let _bytes_written = self
            .handle
            .write_interrupt(self.endpoint_out, &payload, ALLOWED_TIMEOUT)
            .map_err(Error::Rusb)?;

        Ok(())
    }

    pub fn reset_rumble(&self) -> Result<(), Error> {
        self.set_rumble([Rumble::Off, Rumble::Off, Rumble::Off, Rumble::Off])
    }

    fn read_payload(&self) -> Result<[u8; PAYLOAD_SIZE], Error> {
        let mut payload = [0; PAYLOAD_SIZE];

        let bytes_read = self
            .handle
            .read_interrupt(self.endpoint_in, &mut payload, ALLOWED_TIMEOUT)
            .map_err(Error::Rusb)?;

        if bytes_read != PAYLOAD_SIZE || payload[0] != LIBUSB_DT_HID {
            return Err(Error::Payload);
        }

        Ok(payload)
    }

    /// Returns the in and out endpoints for the adapter.
    ///
    /// The in endpoint is for reading controller inputs while the out endpoint
    /// is for writing rumble data.
    fn endpoints(handle: &DeviceHandle<GlobalContext>) -> Result<(u8, u8), Error> {
        let mut endpoint_in = 0;
        let mut endpoint_out = 0;
        let device = handle.device();
        let config = device.config_descriptor(0).map_err(Error::Rusb)?;

        for interface_container in config.interfaces() {
            for interface in interface_container.descriptors() {
                for endpoint in interface.endpoint_descriptors() {
                    match endpoint.direction() {
                        Direction::In => {
                            endpoint_in = endpoint.address();
                        }
                        Direction::Out => {
                            endpoint_out = endpoint.address();
                        }
                    }
                }
            }
        }

        Ok((endpoint_in, endpoint_out))
    }
}

impl Drop for Adapter {
    fn drop(&mut self) {
        let _ = self.close();
    }
}

#[derive(Copy, Clone)]
struct StickRange {
    pub center_x: u8,
    pub center_y: u8,
    pub radius: u8,
}

impl StickRange {
    pub const fn new(center_x: u8, center_y: u8, radius: u8) -> StickRange {
        StickRange {
            center_x,
            center_y,
            radius,
        }
    }

    pub fn restrict(self, x: i16, y: i16) -> (u8, u8) {
        let (center_x, center_y, radius) = (
            self.center_x as i16,
            self.center_y as i16,
            self.radius as i16,
        );

        let xx = Self::clamp(x, center_x - radius, center_x + radius);
        let yy = Self::clamp(y, center_y - radius, center_y + radius);

        (xx as u8, yy as u8)
    }

    fn clamp(n: i16, min: i16, max: i16) -> i16 {
        assert!(min <= max);
        if n < min {
            min
        } else if n > max {
            max
        } else {
            n
        }
    }
}

// TODO: Maybe rename to Calibration and add in triggers?
struct Drift {
    stick_x: i16,
    stick_y: i16,
    substick_x: i16,
    substick_y: i16,
}

impl Drift {
    pub fn new(initial: &Input) -> Drift {
        Drift {
            stick_x: i16::from(MAIN_STICK.center_x) - i16::from(initial.stick_x),
            stick_y: i16::from(MAIN_STICK.center_y) - i16::from(initial.stick_y),
            substick_x: i16::from(C_STICK.center_x) - i16::from(initial.substick_x),
            substick_y: i16::from(C_STICK.center_y) - i16::from(initial.substick_y),
        }
    }

    pub fn correct(&self, mut input: Input) -> Input {
        let (stick_x, stick_y) = MAIN_STICK.restrict(
            i16::from(input.stick_x) + self.stick_x,
            i16::from(input.stick_y) + self.stick_y,
        );

        let (substick_x, substick_y) = C_STICK.restrict(
            i16::from(input.substick_x) + self.substick_x,
            i16::from(input.substick_y) + self.substick_y,
        );

        input.stick_x = stick_x;
        input.stick_y = stick_y;
        input.substick_x = substick_x;
        input.substick_y = substick_y;

        input
    }
}

#[derive(Debug, Copy, Clone)]
pub struct Input {
    pub button_a: bool,
    pub button_b: bool,
    pub button_x: bool,
    pub button_y: bool,

    pub button_left: bool,
    pub button_right: bool,
    pub button_down: bool,
    pub button_up: bool,

    pub button_start: bool,
    pub button_z: bool,
    pub button_r: bool,
    pub button_l: bool,

    pub stick_x: u8,
    pub stick_y: u8,
    pub substick_x: u8,
    pub substick_y: u8,
    pub trigger_left: u8,
    pub trigger_right: u8,
}

impl Input {
    fn parse(payload: &[u8; PAYLOAD_SIZE], port: Port) -> Input {
        let chan = port.channel();
        let b1 = payload[1 + (9 * chan) + 1];
        let b2 = payload[1 + (9 * chan) + 2];

        Input {
            button_a: (b1 & (1 << 0)) != 0,
            button_b: (b1 & (1 << 1)) != 0,
            button_x: (b1 & (1 << 2)) != 0,
            button_y: (b1 & (1 << 3)) != 0,

            button_left: (b1 & (1 << 4)) != 0,
            button_right: (b1 & (1 << 5)) != 0,
            button_down: (b1 & (1 << 6)) != 0,
            button_up: (b1 & (1 << 7)) != 0,

            button_start: (b2 & (1 << 0)) != 0,
            button_z: (b2 & (1 << 1)) != 0,
            button_r: (b2 & (1 << 2)) != 0,
            button_l: (b2 & (1 << 3)) != 0,

            stick_x: payload[1 + (9 * chan) + 3],
            stick_y: payload[1 + (9 * chan) + 4],
            substick_x: payload[1 + (9 * chan) + 5],
            substick_y: payload[1 + (9 * chan) + 6],
            trigger_left: payload[1 + (9 * chan) + 7],
            trigger_right: payload[1 + (9 * chan) + 8],
        }
    }
}

impl Default for Input {
    fn default() -> Input {
        Input {
            button_a: false,
            button_b: false,
            button_x: false,
            button_y: false,

            button_left: false,
            button_right: false,
            button_down: false,
            button_up: false,

            button_start: false,
            button_z: false,
            button_r: false,
            button_l: false,

            stick_x: MAIN_STICK.center_x,
            stick_y: MAIN_STICK.center_y,
            substick_x: C_STICK.center_x,
            substick_y: C_STICK.center_y,
            trigger_left: 0,
            trigger_right: 0,
        }
    }
}

File: src/adapter.zig
const std = @import("std");
const c = @import("c.zig");
const usb = @import("usb.zig");

const payload_len = 37;

pub const Error = error{Payload} || usb.Error;

pub const Adapter = struct {
    const gc_vid = 0x057E;
    const gc_pid = 0x0337;

    const allowed_timeout_ms = 16;

    const Endpoints = struct {
        in: u8,
        out: u8,
    };

    handle: usb.DeviceHandle,
    endpoints: Endpoints,
    calibrations: [4]?Calibration,

    pub fn init(ctx: *usb.Context) Error!Adapter {
        var handle = try ctx.openDeviceWithVidPid(gc_vid, gc_pid);

        try handle.claimInterface(0);

        const endpoints = try findEndpoints(handle);

        // From Dolphin:
        // This call makes Nyko-brand (and perhaps other) adapters work.
        // However it returns LIBUSB_ERROR_PIPE with Mayflash adapters.
        _ = handle.writeControl(0x21, 11, 0x0001, 0, &[_]u8{}, std.time.ms_per_s) catch {};

        // Not sure what this does but Dolphin does it
        _ = handle.writeInterrupt(endpoints.out, &[_]u8{0x13}, allowed_timeout_ms) catch {};

        return Adapter{
            .handle = handle,
            .endpoints = endpoints,
            .calibrations = [_]?Calibration{null} ** 4,
        };
    }

    pub fn deinit(self: Adapter) void {
        self.handle.deinit();
    }

    pub fn readInputs(self: *Adapter) Error![4]?Input {
        const payload = try self.readPayload();
        var inputs = [_]?Input{null} ** 4;

        for (Port.all()) |port| {
            const chan = port.channel();
            // type is 0 if no controller is plugged, 1 if wired, and 2 if wireless
            const controller_type = payload[1 + (9 * chan)] >> 4;
            const connected = controller_type != 0;

            if (!connected) {
                self.calibrations[chan] = null;
                continue;
            }

            const raw = Input.fromPayload(&payload, port);

            if (self.calibrations[chan]) |calibration| {
                inputs[chan] = calibration.correct(raw);
            } else {
                const calibration = Calibration.init(raw);
                inputs[chan] = calibration.correct(raw);
                self.calibrations[chan] = calibration;
            }
        }

        return inputs;
    }

    fn findEndpoints(handle: usb.DeviceHandle) Error!Endpoints {
        const device = handle.device();
        defer device.deinit();

        const config = try device.configDescriptor(0);
        defer config.deinit();

        var in: u8 = 0;
        var out: u8 = 0;

        var interfaces = config.interfaces();
        while (interfaces.next()) |iface| {
            var descriptors = iface.descriptors();
            while (descriptors.next()) |descriptor| {
                var endpoints = descriptor.endpointDescriptors();
                while (endpoints.next()) |endpoint| {
                    switch (endpoint.direction()) {
                        usb.Direction.In => {
                            in = endpoint.address();
                        },
                        usb.Direction.Out => {
                            out = endpoint.address();
                        },
                    }
                }
            }
        }

        return Endpoints{ .in = in, .out = out };
    }

    fn readPayload(self: Adapter) Error![payload_len]u8 {
        var payload: [payload_len]u8 = undefined;

        const bytes_read = try self.handle.readInterrupt(
            self.endpoints.in,
            &payload,
            allowed_timeout_ms,
        );

        if (bytes_read != payload_len or payload[0] != c.LIBUSB_DT_HID) {
            return Error.Payload;
        }

        return payload;
    }
};

pub const Port = enum {
    One,
    Two,
    Three,
    Four,

    pub fn channel(self: Port) usize {
        return switch (self) {
            .One => 0,
            .Two => 1,
            .Three => 2,
            .Four => 3,
        };
    }

    pub fn all() [4]Port {
        return [_]Port{ .One, .Two, .Three, .Four };
    }
};

const Vec2 = struct {
    x: u8,
    y: u8,
};

const StickRange = struct {
    center: u8,
    radius: u8,

    pub fn restrict(self: StickRange, x: i9, y: i9) Vec2 {
        const center = @as(i9, self.center);
        const radius = @as(i9, self.radius);

        const xx = std.math.clamp(x, center - radius, center + radius);
        const yy = std.math.clamp(y, center - radius, center + radius);

        return Vec2{
            .x = std.math.cast(u8, xx) catch unreachable,
            .y = std.math.cast(u8, yy) catch unreachable,
        };
    }
};

const AnalogRange = struct {
    min: u8,
    max: u8,

    pub fn restrict(self: AnalogRange, n: i9) u8 {
        const nn = std.math.clamp(n, @as(i9, self.min), @as(i9, self.min));

        return std.math.cast(u8, nn) catch unreachable;
    }
};

pub const Calibration = struct {
    const main_stick = StickRange{ .center = 0x80, .radius = 0x7F };
    const c_stick = StickRange{ .center = 0x80, .radius = 0x7F };
    const trigger_range = AnalogRange{ .min = 0x00, .max = 0xFF };

    stick_x: i9,
    stick_y: i9,
    substick_x: i9,
    substick_y: i9,
    trigger_left: i9,
    trigger_right: i9,

    pub fn init(initial: Input) Calibration {
        return Calibration{
            .stick_x = @as(i9, main_stick.center) - @as(i9, initial.stick_x),
            .stick_y = @as(i9, main_stick.center) - @as(i9, initial.stick_y),
            .substick_x = @as(i9, c_stick.center) - @as(i9, initial.substick_x),
            .substick_y = @as(i9, c_stick.center) - @as(i9, initial.substick_y),
            .trigger_left = @as(i9, trigger_range.min) - @as(i9, initial.trigger_left),
            .trigger_right = @as(i9, trigger_range.min) - @as(i9, initial.trigger_right),
        };
    }

    pub fn correct(self: Calibration, input: Input) Input {
        var in = input;

        const stick = main_stick.restrict(
            @as(i9, in.stick_x) + self.stick_x,
            @as(i9, in.stick_y) + self.stick_y,
        );

        const substick = c_stick.restrict(
            @as(i9, in.substick_x) + self.substick_x,
            @as(i9, in.substick_y) + self.substick_y,
        );

        in.stick_x = stick.x;
        in.stick_y = stick.y;
        in.substick_x = substick.x;
        in.substick_y = substick.y;
        in.trigger_left = trigger_range.restrict(@as(i9, in.trigger_left) + self.trigger_left);
        in.trigger_right = trigger_range.restrict(@as(i9, in.trigger_right) + self.trigger_right);

        return in;
    }
};

pub const Input = struct {
    button_a: bool,
    button_b: bool,
    button_x: bool,
    button_y: bool,

    button_left: bool,
    button_right: bool,
    button_down: bool,
    button_up: bool,

    button_start: bool,
    button_z: bool,
    button_r: bool,
    button_l: bool,

    stick_x: u8,
    stick_y: u8,
    substick_x: u8,
    substick_y: u8,
    trigger_left: u8,
    trigger_right: u8,

    fn fromPayload(payload: *const [payload_len]u8, port: Port) Input {
        const chan = port.channel();
        const b1 = payload[1 + (9 * chan) + 1];
        const b2 = payload[1 + (9 * chan) + 2];

        return Input{
            .button_a = (b1 & (1 << 0)) != 0,
            .button_b = (b1 & (1 << 1)) != 0,
            .button_x = (b1 & (1 << 2)) != 0,
            .button_y = (b1 & (1 << 3)) != 0,

            .button_left = (b1 & (1 << 4)) != 0,
            .button_right = (b1 & (1 << 5)) != 0,
            .button_down = (b1 & (1 << 6)) != 0,
            .button_up = (b1 & (1 << 7)) != 0,

            .button_start = (b2 & (1 << 0)) != 0,
            .button_z = (b2 & (1 << 1)) != 0,
            .button_r = (b2 & (1 << 2)) != 0,
            .button_l = (b2 & (1 << 3)) != 0,

            .stick_x = payload[1 + (9 * chan) + 3],
            .stick_y = payload[1 + (9 * chan) + 4],
            .substick_x = payload[1 + (9 * chan) + 5],
            .substick_y = payload[1 + (9 * chan) + 6],
            .trigger_left = payload[1 + (9 * chan) + 7],
            .trigger_right = payload[1 + (9 * chan) + 8],
        };
    }
};
